File: II - Linguaggi di programmazione e laboratorio/Esame/Java/29_compito_java_2019_04_11b/GestioneCantina.java
import java.io.*;
import java.util.*;

public class GestioneCantina {

	static List<Impiegato> impiegati = new LinkedList<Impiegato>();
	static List<Visitatore> visitatori = new LinkedList<Visitatore>();
	static Map<Integer, Impiegato> codImp = new HashMap<Integer, Impiegato>();
	static Map<Integer, Integer> codNServ = new HashMap<Integer, Integer>();

	public static void main(String[] args) {

		// PUNTO 1: memorizzazione progetti
		try {
			BufferedReader br = new BufferedReader(new FileReader("impiegati.txt"));
			String line = br.readLine();
			while (line != null) {
				StringTokenizer tok = new StringTokenizer(line);
				int cod = Integer.parseInt(tok.nextToken());
				String tipo = tok.nextToken();
				String nome = br.readLine();
				line = br.readLine();
				if (tipo.equals("sommelier")) {
					tok = new StringTokenizer(line);
					int ore = Integer.parseInt(tok.nextToken());
					double costo = Double.parseDouble(tok.nextToken());
					line = br.readLine();
					String spec = line;
					Impiegato d = new Sommelier(cod, nome, ore, spec, costo);
					impiegati.add(d);
					codImp.put(cod, d);
					codNServ.put(cod, 0);
				} else {
					tok = new StringTokenizer(line);
					int tel = Integer.parseInt(tok.nextToken());
					boolean medico = Boolean.parseBoolean(tok.nextToken());
					int app = Integer.parseInt(tok.nextToken());
					double costo = Double.parseDouble(tok.nextToken());
					Impiegato d = new Guida(cod, nome, tel, medico, app, costo);
					impiegati.add(d);
					codImp.put(cod, d);
					codNServ.put(cod, 0);
				}
				line = br.readLine();
			}
			br.close();
		} catch (IOException e) {
			System.err.println(e);
		} catch (Exception e) {
			System.err.println(e);
		}
		// PUNTO 2: lettura di tutti i visitatori

		try {
			BufferedReader br = new BufferedReader(new FileReader("visitatori.txt"));
			String line = br.readLine();
			while (line != null) {
				int codcli = Integer.parseInt(line);
				String nomeCog = br.readLine();
				Visitatore c = new Visitatore(codcli, nomeCog);
				visitatori.add(c);
				line = br.readLine();
				while (line != null && !line.equals("")) {
					StringTokenizer tok = new StringTokenizer(line);
					int cod = Integer.parseInt(tok.nextToken());
					double ore = Double.parseDouble(tok.nextToken());
					Impiegato dip = codImp.get(cod);
					c.addServizio(dip, ore);
					int current = codNServ.get(cod);
					codNServ.put(cod, current + 1);
					line = br.readLine();
				}
				line = br.readLine();
			}
			br.close();
		} catch (IOException e) {
			System.err.println(e);
		}

		// PUNTO 3: stampa di tutti i impiegati
		System.out
				.println("nome, codice, tipo, ore settimanali, specialità, telefono, inglese, app.sett., costo orario");
		for (Impiegato d : impiegati)
			System.out.println(d);
		System.out.println();
		// //PUNTO 4: stampa visitatori
		for (Visitatore c : visitatori)
			System.out.println(c);
		System.out.println();

		// PUNTO 5: stampe del Impiegato che ha fatto più servizi
		int cod = Integer.parseInt(args[0]);
		System.out.println(codNServ.get(cod));

	}
}

File: II - Linguaggi di programmazione e laboratorio/Esame/Python/29b_compito_java_20190411/GestioneCantina.py
from Impiegato import Impiegato
from Guida import Guida
from Sommelier import Sommelier
from Visitatore import Visitatore
from Servizio import Servizio

import sys


class GestioneCantina:

    def main():

        impiegati = []
        visitatori = []
        cod_to_impiegato = {}

        try:
            f = open("impiegati.txt", "r")
            line = f.readline().strip()

            while line != "":
                tok = line.split()
                codice = int(tok[0])
                tipo = tok[1]

                nome = f.readline().strip()

                line = f.readline().strip()
                tok = line.split()
                telefono_ore = int(tok[0])

                if tipo == "guida":
                    inglese = bool(tok[1])
                    app_settimanali = int(tok[2])
                    costo = float(tok[3])

                    g = Guida(codice, tipo, nome, telefono_ore, inglese, app_settimanali, costo)
                    impiegati.append(g)
                    cod_to_impiegato[codice] = g

                else:
                    costo = float(tok[1])
                    specialita = f.readline().strip()

                    s = Sommelier(codice, tipo, nome, telefono_ore, costo, specialita)
                    impiegati.append(s)
                    cod_to_impiegato[codice] = s

                line = f.readline().strip()

        except IOError as i:
            print(i)
        except Exception as e:
            print(e)

        try:
            f = open("visitatori.txt", "r")
            line = f.readline().strip()

            while line != "":
                codice_visitatore = int(line)
                nome_visitatore = f.readline().strip()

                vis = Visitatore(codice_visitatore, nome_visitatore)
                visitatori.append(vis)

                line = f.readline().strip()

                while line != "":
                    tok = line.split()
                    codice_impiegato = int(tok[0])
                    ore_servizio = float(tok[1])

                    dip = cod_to_impiegato[codice_impiegato]

                    ser = Servizio(codice_impiegato, ore_servizio, dip.get_nome(), dip.get_costo())
                    vis.add_servizio(ser)
                    dip.incrementa_numero_servizi_svolti()

                    line = f.readline().strip()

                line = f.readline().strip()

        except IOError as i:
            print(i)
        except Exception as e:
            print(e)

        print("nome, codice, tipo, ore settimanali, specialità, telefono, inglese, app.sett., costo orario")

        for i in impiegati:
            print(i)
        print()

        for v in visitatori:
            print(v)
        print()

        try:
            codice_inserito = int(sys.argv[1])

            for i in impiegati:
                if codice_inserito == i.get_codice():
                    print("Numero di servizi svolti da " + i.get_nome() + ": " + str(i.get_numero_servizi()))

        except IOError as i:
            print(i)
        except Exception as e:
            print(e)


if __name__ == "__main__":
    GestioneCantina.main()
